### PR TITLE
Update and lock down deps

### PR DESCRIPTION
**Pull request by wesleytodd**
The [recent security issue](https://github.com/dominictarr/event-stream/issues/116) made me reconsider the approach of leaving "unlocked" dependencies on this project.  In light of that I updated, audited and locked down the dependencies.  I will leave this open for a day for comments or concerns, but if none arise I would like to move forward with this.

EDIT: If you react to this but do not post reasonable concerns or arguments against I will not be taking you into consideration.  A reaction is just not enough to know your thoughts.

### CODE REVIEW

**Comment by LinusU** [comment]
I personally don't like this approach because of a few reasons:

1) **It doesn't actually fix the problem, it just (might, see point 3) mitigate it a bit.** All the dependencies that we depend on probably have `^`-dependencies, and thus this will not prevent them being updated.

2) **It can prevent security fixes from reaching end users.** If any of our locked down dependencies releases a security fix, we need to be on top of that right away and release a new package in order for that to reach our users. This creates unnecessary delays in getting patches out to users, especially if this happens a few hops down the line.

3) **It places a burden on us, the maintainers, of this package.** Will we go thru and check the code of our dependencies every time we bump them? Because if we don't, and just blindly bump them as usual, we haven't actually mitigated the problem at all...

4) **It reduces the ability for npm to dedupe dependencies.** If every package specifies hard requirements, the chance of them being out of sync with each other is very high. This results in multiple versions of the same package being used at once, which not only affects network and disk usage, but also increases RAM and CPU usage since the package will need to be loaded multiple times as well.

5) **There are better solutions to this problem.** There already exists a great solution to lock down dependencies, the `package-lock.json` file. Contrary to what some people seem to think (not anyone specifically in this thread, but I've seen this a lot), versions never get bumped automatically in this file. E.g. if you have dependency `a` and `b`, and `a` has a dependency on `b`, bumping `a` will not pull the latest `b`. It's also very easy to upgrade specific dependencies that are not in your top-level dependencies. E.g. `npm update foobar --depth=99`.

    Using the `package-lock.json` you can build tooling that helps you set the level of safety you want. If you want to audit every change that goes into your applications (which you probably should) then the best way to that is to have tooling that makes a PR whenever any of your dependencies, *including transient dependencies,* change that just bumps that version in the `package-lock.json` file (by running `npm update package-name --depth=99`). You can then audit that specific change by seeing what's changed in that library, and decide if you want to merge or not.

**tl;dr:** there is no way to get security for free, I don't see how this change here will help anyone being more secure

**Comment by wesleytodd** [comment]
Great points @LinusU, this is the productive conversation I was hoping to start.  I have a few things on my plate today so might not be able to do a full response to all of these points and for sure wouldn't have been able to merge/publish, but thanks for taking the time to write great detailed counter points so quickly!

**Comment by wesleytodd** [comment]
> It doesn't actually fix the problem, it just (might, see point 3) mitigate it a bit.

Agreed, this is no perfect fix.  This is strictly an exposure minimization effort.  My hope is that leading by example will help maintainers of other popular modules, like the ones we depend on, to follow.  

> It can prevent security fixes from reaching end users.

This is a very legitimate concern with this approach.  I think that being that this is a tool you run in development, the types of security concerns *we* have are worse from a rogue update than from a runtime vulnerability.  Similar to the compromise of `event-stream`, I think the most concerning vector of attack for this module is at install time, pulling in un-trusted code.  So IMO, this is the correct trade off specifically for this type of module.

> It places a burden on us, the maintainers of this package.

Yes, and I am fine with this.  If it becomes too much of a burden I was thinking of adding greenkeeper to do automated PR's.  Then if the bot opens the PR, all I have to do is compare some diffs in the upstream packages.  Yes I could miss something, and yes this is extra work, but it is better than nothing like we have today.

> It reduces the ability for npm to dedupe dependencies.

Again, this is true, but I think it is a tradeoff.  Being a dev tool with a relatively small dependency tree, I think this is a reasonable sacrifice for improved end user security.

> There are better solutions to this problem.

I am on the fence on this.  I think putting security in the users hands is a great way to have them all mess it up.  Very few developers I know follow even the most simple security review for their node modules.  Package lock is great for those who use it correctly, but most people I know just blow away their lock when they are updating.  This is because module authors have not been good about compatibility, so sometimes it just doesn't work and you don't know why.  Ror the average developer, that means the most productive option is to `rm package-lock.json && npm i`.  It is sad but true.

I think we, the module maintainers, are in the best place to take care of the user base; and if it means extra work for us where the benefit is not having massive security holes for the whole community, then so be it.  Again, I am not entirely happy about this, and I wish users were more thoughtful of security, and that the tooling was better.  But in absence of all of that, this is one little thing that we as maintainers can do to make it better.

Lastly, I am amazed at the strength with which most users fight against locking deps.  I have heard many great arguments, but none which inspire the zeal I hear from people on the topic.  Either way, I agree it is a trade off and I wanted to bring up the issue rather than just sit idly by as public incidents occur which effect the legitimacy of the entire npm module ecosystem.

**Comment by wesleytodd** [comment]
Ok, this has sat for a while. There are risks, but it is nothing more than the rest of the node community has already accepted.  No one else came out in defense of the idea, and @LinusU made great points against.  I will close this out.
